core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1535,6 +1545,43 @@ class KafkaController(val config: KafkaConfig, zkClient: KafkaZkClient, time: Ti
 
 }
 
+case class ReassignmentStep(currentReplicas: Seq[Int], drop: Seq[Int], add: Option[Int], targetReplicas: Seq[Int]) {
+  override def toString: String = {
+    val currentFragment = currentReplicas.mkString("current replicas=", ","," ")
+    val dropFragment = if (drop.isEmpty) "" else drop.mkString("-", ","," ")
+    val addFragment = if (add.isEmpty) "" else s"+${add.get} "
+    val targetFragment = targetReplicas.mkString("target replicas=", ",", "")
+    s"$currentFragment$dropFragment$addFragment$targetFragment"
+  }
+}
+
+private[controller] object ReassignmentHelper {
+
+  // Rules
+  // - keep current leader as long as possible to avoid unnecessary leader elections.
```

Comment:
Hmm... I think we'd want to change to the new preferred leader as soon as possible. Often times we are doing a rebalance in order to relieve stress from the current replicas and this would move client load away from the current leader. We'd definitely want to bring the new preferred leader into sync first (which looks like what we're doing below).

Reply:
It makes sense. We chose this way to avoid unnecessary leader elections by doing it once at the end when everything else but that is in place. On the other hand though it makes sense what you're saying and that also requires only one leader election.

Reply:
One thought here. There is a case when the new preferred leader is not online which complicates electing the new leader. I suggest that the algorithm should try adding the new preferred leader first and electing it, but if it's not online then elect a temporary leader and reelect as soon as it brought the new preferred leader online. Although this may cause some churn in the cluster, in particular if you have a lot of clients, from the broker's perspective it might be better to relieve them from stress.